### PR TITLE
[RISCV] Add CORE-V headers

### LINKED ISSUE #74: CORE-V C API Specification
Is this defined? See the comment at https://reviews.llvm.org/D157510#4573494

### PR DESCRIPTION
**Pull request by melonedo**
Add header files for all CORE-V extensions that have C builtin functions following [D155647](https://reviews.llvm.org/D155647). The definitions of existing builtin functions are also adjusted accordingly.

Note that [D155647](https://reviews.llvm.org/D155647) defines new C functions without prefix `__builtin` (see below, also refer to [RISCV-C-API Specification#Intrinsic Functions](https://github.com/riscv-non-isa/riscv-c-api-doc/blob/master/riscv-c-api.md#intrinsic-functions)), while this PR simply re-declares existing builtins.

```c
static __inline__ uint32_t __attribute__((__always_inline__, __nodebug__))
__riscv_rol_32(uint32_t __x, uint32_t __y) {
  return __builtin_rotateleft32(__x, __y);
}
```

The header files are extracted from the specification with this script:

<details>
<summary>script</summary>

```python
import re
from pathlib import Path
from textwrap import dedent

# identify target-independent builtin functions
NAME_TO_EXTENSION_MAP = dict(abs="alu")

with open("corev-builtin-spec.md", "rt") as file:
    spec = file.read()
exts = dict()

# rely on "####" as an indication of a builtin
for f in re.finditer(r"#### `(.+)`", spec):
    r = re.match(
        r"\w+ __builtin_((?:riscv_cv_([^_]+))?\w+)\s*\(.+\)",
        f[1],
    )
    if r is None:
        print("Failed to parse:", f[1])
        continue
    name = r[1]
    ext = NAME_TO_EXTENSION_MAP.get(name, r[2])
    # print(name, ext)
    exts.setdefault(ext, []).append(f[1])

include = Path("include")
include.mkdir(exist_ok=True)

for e, funcs in exts.items():
    prologue = dedent(f"""\
    #ifndef __RISCV_COREV_{e.upper()}_H
    #define __RISCV_COREV_{e.upper()}_H

    #include <stdint.h>

    #if defined(__cplusplus)
    extern "C" {{
    #endif

    #if defined(__riscv_xcv{e})
    """)
    epilogue = dedent(f"""\
    #endif // defined(__riscv_xcv{e})

    #if defined(__cplusplus)
    }}
    #endif

    #endif // define __RISCV_COREV_{e.upper()}_H
    """)
    ext_header = include / f"riscv_corev_{e}.h"
    with ext_header.open("wt") as file:
        file.write(prologue)
        for func in funcs:
            # file.write(
            #     "\nstatic __inline__ uint64_t __attribute__((__always_inline__, __nodebug__))\n"
            # )
            file.write(func)
            file.write(";\n")
        file.write(epilogue)

```

</details>

Fix #74.
CC @MaryBennett, @PaoloS02.

### CODE REVIEW

**Comment by melonedo** [comment]
The failure on Mac happens in some python scripts and should be irrelevant, see https://reviews.llvm.org/D41784.

File: clang/lib/Headers/riscv_corev_alu.h [hunk 1]
```diff
@@ -0,0 +1,38 @@
+#ifndef __RISCV_COREV_ALU_H
```
Comment (PaoloS02) [inline clang/lib/Headers/riscv_corev_alu.h:1]: Licence?

File: clang/include/clang/Basic/BuiltinsRISCVCOREV.def [hunk 2]
```diff
@@ -84,26 +84,26 @@ TARGET_BUILTIN(simd_dotup_h, "UZiUZiUZi", "nc", "xcvsimd")
 TARGET_BUILTIN(simd_dotup_b, "UZiUZiUZi", "nc", "xcvsimd")
 TARGET_BUILTIN(simd_dotup_sc_h, "UZiUZiUs", "nc", "xcvsimd")
 TARGET_BUILTIN(simd_dotup_sc_b, "UZiUZiUc", "nc", "xcvsimd")
-TARGET_BUILTIN(simd_dotusp_h, "ZiUZiSZi", "nc", "xcvsimd")
-TARGET_BUILTIN(simd_dotusp_b, "ZiUZiSZi", "nc", "xcvsimd")
+TARGET_BUILTIN(simd_dotusp_h, "ZiUZiUZi", "nc", "xcvsimd")
+TARGET_BUILTIN(simd_dotusp_b, "ZiUZiUZi", "nc", "xcvsimd")
```
Comment (PaoloS02) [inline clang/include/clang/Basic/BuiltinsRISCVCOREV.def:88]: I might be missing something, but the specs here: https://github.com/openhwgroup/cv32e40p/blob/62bec66b36182215e18c9cf10f723567e23878e9/docs/source/instruction_set_extensions.rst talk about the second operand being treated as signed ("Note: rs1 is treated as unsigned, while op2 is treated as signed."). Am I getting this wrong?

File: clang/include/clang/Basic/BuiltinsRISCVCOREV.def [hunk 3]
```diff
@@ -84,26 +84,26 @@ TARGET_BUILTIN(simd_dotup_h, "UZiUZiUZi", "nc", "xcvsimd")
...
 TARGET_BUILTIN(simd_dotusp_sc_h, "ZiUZiSs", "nc", "xcvsimd")
 TARGET_BUILTIN(simd_dotusp_sc_b, "ZiUZiSc", "nc", "xcvsimd")
-TARGET_BUILTIN(simd_dotsp_h, "ZiSZiSZi", "nc", "xcvsimd")
-TARGET_BUILTIN(simd_dotsp_b, "ZiSZiSZi", "nc", "xcvsimd")
-TARGET_BUILTIN(simd_dotsp_sc_h, "ZiSZiSs", "nc", "xcvsimd")
-TARGET_BUILTIN(simd_dotsp_sc_b, "ZiSZiSc", "nc", "xcvsimd")
+TARGET_BUILTIN(simd_dotsp_h, "ZiUZiUZi", "nc", "xcvsimd")
+TARGET_BUILTIN(simd_dotsp_b, "ZiUZiUZi", "nc", "xcvsimd")
+TARGET_BUILTIN(simd_dotsp_sc_h, "ZiUZiSs", "nc", "xcvsimd")
+TARGET_BUILTIN(simd_dotsp_sc_b, "ZiUZiSc", "nc", "xcvsimd")
```
Comment (PaoloS02) [inline clang/include/clang/Basic/BuiltinsRISCVCOREV.def:94]: If my previous comment makes sense here all operands should be signed? Or am I reading wrong the operand specifiers for these simd instructions?

**Comment by ChunyuLiao** [comment]
funan back to school , I think we can close this.

File: clang/include/clang/Basic/BuiltinsRISCVCOREV.def (hunk 2 above)
Comment (PaoloS02) [reply clang/include/clang/Basic/BuiltinsRISCVCOREV.def:88]: Double checked the documentation. It's actually correct.

File: clang/include/clang/Basic/BuiltinsRISCVCOREV.def (hunk 3 above)
Comment (PaoloS02) [reply clang/include/clang/Basic/BuiltinsRISCVCOREV.def:94]: Same as above.

**Comment by PaoloS02** [comment]
Implemented in new patches.

**Pull request closed without merging**